artiq/applets/big_number.py

```diff
@@ -26,47 +27,78 @@ def __init__(self, args, ctl):
         QtWidgets.QStackedWidget.__init__(self)
         self.dataset_name = args.dataset
         self.ctl = ctl
+        self.value = 0
+        self.metadata = {}
 
-        self.lcd_widget = QResponsiveLCDNumber()
-        self.lcd_widget.setDigitCount(args.digit_count)
-        self.lcd_widget.doubleClicked.connect(self.start_edit)
-        self.addWidget(self.lcd_widget)
+        self.display_widget = QResponsiveLabel()
+        self.display_widget.setAlignment(
+            QtCore.Qt.AlignRight | QtCore.Qt.AlignVCenter)
+        self.display_widget.doubleClicked.connect(self.start_edit)
+        self.addWidget(self.display_widget)
 
         self.edit_widget = QCancellableLineEdit()
-        self.edit_widget.setValidator(QtGui.QDoubleValidator())
-        self.edit_widget.setAlignment(QtCore.Qt.AlignRight)
+        self.edit_widget.setAlignment(
+            QtCore.Qt.AlignRight | QtCore.Qt.AlignVCenter)
         self.edit_widget.editCancelled.connect(self.cancel_edit)
         self.edit_widget.returnPressed.connect(self.confirm_edit)
         self.addWidget(self.edit_widget)
 
         font = QtGui.QFont()
         font.setPointSize(60)
+        self.display_widget.setFont(font)
         self.edit_widget.setFont(font)
 
-        self.setCurrentWidget(self.lcd_widget)
+        self.setCurrentWidget(self.display_widget)
+
+    def get_unscaled_format(self):
+        scale = self.metadata.get("scale", None)
+        unit = self.metadata.get("unit", "")
+        unit_scale = scale_from_metadata({"unit": unit})
+        if scale is not None:
+            value = self.value/scale
+        else:
+            value = self.value/unit_scale
+        return f"{value} {unit}".strip()
 
     def start_edit(self):
-        # QLCDNumber value property contains the value of zero
-        # if the displayed value is not a number.
-        self.edit_widget.setText(str(self.lcd_widget.value()))
+        if self.value == 0:
+            self.edit_widget.setText("---")
+        else:
+            self.edit_widget.setText(self.get_unscaled_format())
         self.edit_widget.selectAll()
         self.edit_widget.setFocus()
         self.setCurrentWidget(self.edit_widget)
 
     def confirm_edit(self):
-        value = float(self.edit_widget.text())
-        self.ctl.set_dataset(self.dataset_name, value)
-        self.setCurrentWidget(self.lcd_widget)
+        text = self.edit_widget.text().split()
+        try:
+            value = float(text[0])
+        except ValueError:
+            self.display_widget.setText("---")
+            self.setCurrentWidget(self.display_widget)
+            return
+        unit = text[1] if len(text) > 1 else ""
+        scale = scale_from_metadata({"unit": unit})
+        value = value*scale
+        precision = self.metadata.get("precision", None)
```

Comment:
see if you can increase the precision if the user inputs more decimal points than previously. Say precision is 2 and the value is 5.44 originally, and then the user changes it to 5.654 in this widget. In this case I would recommend increasing the precision to 3 such that the display mirrors appropriately. Precision is only a display parameter anyways

Reply:
Since the `edit_widget` reflects the raw value of the input (just like the current value in the edit dataset dialog box), changing the precision that way will be a bit odd. For example, if the input value of the dataset is 123.456 and the precision is 2, the output shown will be 123.46 in both dataset and display if big_number. However if we enter the edit mode, we will see the same input value 123.456 stored, and editing this, say 123.4567 will change the precision to 4. Since this adaptive precision is not the current behavior of the dataset, I think it would be fitting to leave the precision setting to the dataset alone for more consistency.

Reply:
agree :+1: